extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProvider.java

```diff
@@ -59,21 +62,56 @@ final class S3SeekableChannelProvider implements SeekableChannelsProvider {
     private final S3AsyncClient s3AsyncClient;
     private final S3Instructions s3Instructions;
 
-    private final Map<URI, Long> uriToFileSize;
+    private SoftReference<KeyedObjectHashMap<URI, FileSizeInfo>> uriToFileSizeSoftRef;
```

Comment:
Note, we need to do something about threading here.

Reply:
```
    @SuppressWarnings({"unused", "rawtypes"})
    private static final AtomicReferenceFieldUpdater<S3SeekableChannelProvider, SoftReference> URI_TO_FILE_SIZE_REF_UPDATER =
            AtomicReferenceFieldUpdater.newUpdater(S3SeekableChannelProvider.class, SoftReference.class, "uriToFileSizeRef");

    private volatile SoftReference<Map<URI, FileSizeInfo>> uriToFileSizeRef;
```

```
    private Map<URI, FileSizeInfo> fileSizeCache() {
        SoftReference<Map<URI, FileSizeInfo>> cacheRef = uriToFileSizeRef;
        Map<URI, FileSizeInfo> cache;
        while ((cacheRef = uriToFileSizeRef) == null || (cache = cacheRef.get()) == null) {
            if (URI_TO_FILE_SIZE_REF_UPDATER.compareAndSet(this, cacheRef,
                    new SoftReference<>(cache = new KeyedObjectHashMap<>(URI_MATCH_KEY)))) {
                return cache;
            }
        }
        return cache;
    }

    /**
     * Update the file size cache with the given URI and size.
     */
    private static void updateFileSizeCache(
            @NotNull final Map<URI, FileSizeInfo> fileSizeCache,
            @NotNull final URI uri,
            final long size) {
```

Note, you don't need to initialize in the constructor if you do it this way, and you should get the map once per loop.